File: lab1/lab1/java/Canvas.java
//
//  Canvas.java
//

import java.awt.*;
import java.nio.*;
import java.awt.event.*;
import javax.media.opengl.*;
import javax.media.opengl.awt.GLCanvas;
import java.util.*;

public class Canvas {

    ///
    // canvas size information
    ///
    private int width;
    private int height;

    ///
    // point-related data
    ///
    private Vector<Float> points;
    private short numElements;
    private Vector<Short> elements;

    ///
    // color-related data
    ///
    private Vector<Float> colors;
    private Color currentColor;

    Canvas( int w, int h )
    {
        width = w;
        height = h;
	numElements = 0;
        points = new Vector<Float>();
        colors = new Vector<Float>();
        elements = new Vector<Short>();
        currentColor = new Color( 0.0f, 0.0f, 0.0f );
    }

    public void clear()
    {
        // easiest method:  just allocate new ones!
	numElements = 0;
        points = new Vector<Float>();
        colors = new Vector<Float>();
        elements = new Vector<Short>();
        currentColor = new Color( 0.0f, 0.0f, 0.0f );
    }

    public void setColor( float r, float g, float b )
    {
        currentColor = new Color( r,g, b );
    }

    public void setPixel( int x, int y )
    {
	points.add( new Float(x) );
	points.add( new Float(y) );
	points.add( new Float(-1.0) );  // fixed Z depth
	points.add( new Float(1.0) );

	float v[] = currentColor.getRGBColorComponents( null );
	colors.add( new Float(v[0]) );
	colors.add( new Float(v[1]) );
	colors.add( new Float(v[2]) );
	colors.add( new Float(1.0 ) );  // alpha channel

        elements.add( new Short(numElements) );
	numElements += 1;
    }

    /**
     * get the vertex points for the current shape
     */
    public Buffer getVertices()
    {
        float v[] = new float[points.size()];
        for( int i=0; i < points.size(); i++ ) {
            v[i] = (points.elementAt(i)).floatValue();
        }
        return FloatBuffer.wrap( v );
    }

    /**
     * get the array of elements for the current shape
     */
    public Buffer getElements()
    {
        short e[] = new short[elements.size()];
        for( int i=0; i < elements.size(); i++ ) {
            e[i] = (elements.elementAt(i)).shortValue();
        }
        return ShortBuffer.wrap( e );
    }

    /**
     * get the colors for the current shape
     */
    public Buffer getColors()
    {
        float v[] = new float[colors.size()];
        for( int i=0; i < colors.size(); i++ ) {
            v[i] = (colors.elementAt(i)).floatValue();
        }
        return FloatBuffer.wrap( v );
    }

    public int nVertices()
    {
        return numElements;
    }

}

File: lab1/lab1/c/Canvas.c
//
//  Canvas.c
//
//  Routines for adding points to create a new mesh.
//
//  Students are not to modify this file.
//

#ifdef __APPLE__
#include <OpenGL/gl.h>
#else
#ifdef WIN32
#include <windows.h>
#endif
#include <GL/glew.h>
#include <GL/gl.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "Canvas.h"
#include "floatVector.h"

///
// Constructor
//
// @param w width of canvas
// @param h height of canvas
///

Canvas *Canvas_create( int w, int h ) {
    Canvas *C = calloc( 1, sizeof(Canvas) );
    if( C != NULL ) {
        C->width = w;
        C->height = h;
        // all other fields contain zeroes
    }
    return( C );
}


///
// Destructor
//
// @param C canvas to destroy
///

void Canvas_destroy( Canvas *C ) {
    if( !C )
        return;
    free( C );
}


///
// clear the canvas
//
// @param C The Canvas to use
///
void Canvas_clear( Canvas *C )
{
    if( C->pointArray ) {
        free( C->pointArray );
        C->pointArray = 0;
    }
    if( C->elemArray ) {
        free( C->elemArray );
        C->elemArray = 0;
    }
    if( C->colorArray ) {
        free( C->colorArray );
        C->colorArray = 0;
    }
    floatVectorClear( &(C->points) );
    floatVectorClear( &(C->colors) );
    C->numElements = 0;
    C->currentColor[0] = 0.0;
    C->currentColor[1] = 0.0;
    C->currentColor[2] = 0.0;
}


///
// change the current drawing color
//
// @param r The red component of the new color (between 0-1)
// @param g The green component of the new color (between 0-1)
// @param b The blue component of the new color (between 0-1);
// @param C The Canvas to use
///
void Canvas_setColor( float r, float g, float b, Canvas *C )
{
    C->currentColor[0] = r;
    C->currentColor[1] = g;
    C->currentColor[2] = b;
}


///
// set a pixel to the current drawing color
//
// @param x The x coord of the pixel to be set
// @param y The y coord of the pixel to be set
// @param C The Canvas to use
///
void Canvas_setPixel( int x0, int y0, Canvas *C )
{
    floatVectorPushBack( &(C->points), (float) x0 );
    floatVectorPushBack( &(C->points), (float) y0 );
    floatVectorPushBack( &(C->points), -1.0 );  // fixed Z depth
    floatVectorPushBack( &(C->points), 1.0 );

    floatVectorPushBack( &(C->colors), C->currentColor[0] );
    floatVectorPushBack( &(C->colors), C->currentColor[1] );
    floatVectorPushBack( &(C->colors), C->currentColor[2] );
    floatVectorPushBack( &(C->colors), 1.0 );   // alpha channel

    C->numElements += 1;
}


///
// gets the array of vertices for the current shape
//
// @param C The Canvas to use
///
float *Canvas_getVertices( Canvas *C )
{
    int i;

    // delete the old point array if we have one
    if( C->pointArray ) {
        free( C->pointArray );
    }

    // create and fill a new point array
    C->pointArray = (float *) malloc(
        floatVectorSize(&(C->points)) * sizeof(float) );
    if( C->pointArray == 0 ) {
    	perror( "point allocation failure" );
	exit( 1 );
    }
    for( i = 0; i < floatVectorSize(&(C->points)); i++ ) {
        C->pointArray[i] = C->points.vec[i];
    }

    return C->pointArray;
}


///
// gets the array of elements for the current shape
//
// @param C The Canvas to use
///
GLushort *Canvas_getElements( Canvas *C )
{
    int i;

    // delete the old point array if we have one
    if( C->elemArray ) {
        free( C->elemArray );
    }

    // create and fill a new point array
    C->elemArray = (GLushort *) malloc(
        floatVectorSize(&(C->points)) * sizeof(GLushort) );
    if( C->elemArray == 0 ) {
    	perror( "element allocation failure" );
	exit( 1 );
    }
    for( i = 0; i < floatVectorSize(&(C->points)); i++ ) {
        C->elemArray[i] = i;
    }

    return C->elemArray;
}


///
// gets the array of colors for the current shape
//
// @param C The Canvas to use
///
float *Canvas_getColors( Canvas *C )
{
    int i;

    // delete the old color array if we have one
    if( C->colorArray ) {
        free( C->colorArray );
    }

    // create and fill a new color array
    C->colorArray = (float *) malloc(
        floatVectorSize(&(C->colors)) * sizeof(float) );
    if( C->colorArray == 0 ) {
    	perror( "color allocation failure" );
	exit( 1 );
    }
    for( i = 0; i < floatVectorSize(&(C->colors)); i++ ) {
        C->colorArray[i] = C->colors.vec[i];
    }

    return C->colorArray;
}


///
// returns number of vertices in current shape
//
// @param C The Canvas to use
///
int Canvas_nVertices( Canvas *C )
{
    return C->numElements;
}
